# STAT 207 Lab Assignment 4 - Group Part - [12 total points] - Solutions

This group lab will also be atypical in that the number of points is larger than the typical 5.

## <u>Case Study</u>: Comparing Movie Ratings based on Famous Actors



### Movie Rating Dataset Information

In the individual and group assignments we will use the data in the `imdb.csv` to pursue our research goals.

Each row in the `imdb.csv` file contains the following information about **POPULAR** `films` as well as `series` listed on IMDB.
* `Name`
* `Date`: the date film/series was released.
* `Rate`: the rating of the film/series
* `Votes`: how many votes this film/series received on IMDB
* `Genre`: a **list** of genres that the film/series is considered to be. Most films/series have multiple genres listed (ex: 'Action, Adventure, Thriller')
* `Duration`: the duration of the film/series
* `Type`: whether the row is a `Film` or `Series`
* `Certificate`: the content rating of the film (ex: 'G', 'PG', 'PG-13', 'R')
* `Nudity`: the amount of nudity in the film/series
* `Violence`: the amount of violence in the film/series
* `Profanity`: the amount of profanity in the film/series
* `Alcohol`: the amount of alcohol in the film/series
* `Frightening`: the amount of frightening content in the film/series

### Movie Actors Dataset Information

In addition, in the group assignments we will use the data in the `ReelView.csv` to pursue our research goals.
Each row in the `ReelView.csv` file contains the following additional information about `films` as well as `series` listed on the ReelView website.

* `title`
* `release_date`
* `director`: the director of the film/series
* `cast`: the list of actors in the film/series




### Group Assignment Goal

In this group assignment you will pick two famous actors and compare the distribution of ratings for films that these actors each apppear in.

<p>&nbsp;</p>
<table style="border: none;border-collapse: collapse;width:102pt;">
    <tbody>
        <tr>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:general;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;width:51pt;">Problem</td>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:general;vertical-align:bottom;border:.5pt solid windowtext;border-left:none;width:51pt;">Points</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.5</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.6</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.7</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.5</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
    </tbody>
</table>




In [150]:
import pandas as pd
import seaborn as sns
sns.set()
import matplotlib.pyplot as plt
import numpy as np

## 1. IMDB Dataset

### 1.1. Reading the csv

1. Read the `imdb.csv` into a dataframe. The strings that are used to represent missing values in this csv file are `No Rate`, `No Votes`, and `None`. Make sure that these values are converted to NaN values when you read the csv file.
2. Show the first 5 rows.
3. Show how many rows this dataframe has.

In [151]:
df = pd.read_csv("imdb.csv", na_values = ['No Rate','No Votes','None'])
df.head()

Name  Date  Rate     Votes  \
0               No Time to Die  2021   7.6  107163.0   
1                   The Guilty  2021   6.3   64375.0   
2    The Many Saints of Newark  2021   6.4   27145.0   
3  Venom: Let There Be Carnage  2021   6.4   30443.0   
4                         Dune  2021   8.3   84636.0   

                         Genre  Duration  Type Certificate    Nudity  \
0  Action, Adventure, Thriller     163.0  Film       PG-13      Mild   
1       Crime, Drama, Thriller      90.0  Film           R       NaN   
2                 Crime, Drama     120.0  Film           R  Moderate   
3    Action, Adventure, Sci-Fi      97.0  Film       PG-13       NaN   
4     Action, Adventure, Drama     155.0  Film       PG-13       NaN   

   Violence Profanity   Alcohol Frightening  
0  Moderate      Mild      Mild    Moderate  
1       NaN    Severe       NaN    Moderate  
2    Severe    Severe  Moderate    Moderate  
3  Moderate  Moderate      Mild    Moderate  
4  Moderate       NaN      Mild    Moderate

In [152]:
print("This dataset has ", df.shape[0], " rows.")

This dataset has  6178  rows.


### 1.2. Missing Value Cleaning

1. First create a dataframe that only contains the the columns that we care about for this pre-analsys: `Name`, `Rate`, `Date`, and `Type`.
2. Then drop all rows with NaN values from this dataframe.
3. How many rows did we drop?

In [153]:
df_four = df[["Name","Rate","Date","Type"]]
df_four

Name  Rate  Date  Type
0                             No Time to Die   7.6  2021  Film
1                                 The Guilty   6.3  2021  Film
2                  The Many Saints of Newark   6.4  2021  Film
3                Venom: Let There Be Carnage   6.4  2021  Film
4                                       Dune   8.3  2021  Film
...                                      ...   ...   ...   ...
6173  The Human Centipede II (Full Sequence)   3.8  2011  Film
6174                        Double Indemnity   8.3  1944  Film
6175      Before the Devil Knows You're Dead   7.3  2007  Film
6176                              Queen Bees   6.0  2021  Film
6177                              Death Race   6.3  2008  Film

[6178 rows x 4 columns]

In [154]:
df_four.dropna()

Name  Rate  Date  Type
0                             No Time to Die   7.6  2021  Film
1                                 The Guilty   6.3  2021  Film
2                  The Many Saints of Newark   6.4  2021  Film
3                Venom: Let There Be Carnage   6.4  2021  Film
4                                       Dune   8.3  2021  Film
...                                      ...   ...   ...   ...
6173  The Human Centipede II (Full Sequence)   3.8  2011  Film
6174                        Double Indemnity   8.3  1944  Film
6175      Before the Devil Knows You're Dead   7.3  2007  Film
6176                              Queen Bees   6.0  2021  Film
6177                              Death Race   6.3  2008  Film

[5993 rows x 4 columns]

### 1.3. Looking for Duplicate Names

#### 1.3.1. Duplicate Example

Create a dataframe that is only comprised of rows with the Name 'Twin Peaks'.



In [155]:
df_duplicate =  df_four.loc[df["Name"] == "Twin Peaks"]
df_duplicate

Name  Rate  Date    Type
605   Twin Peaks   8.8  2017  Series
1380  Twin Peaks   8.5  2017  Series

#### 1.3.2. List of Non-Duplicated Movie Names

Unfortunately, what you'll see in 1.3.1 is that there are multiple 'Twin Peaks' series with the same `Name` (and Date for that matter), but a different set of other values.

It's possible that this series is a **duplicate** and/orrepresents some sort of "entry error" on the part of whoever compiled this dataset. Unfortunately, duplicate or "entry error" rows in a dataframe can complicate our analyses.

To make things easier for our analysis, let's create a list of all movie `Names` in this dataframe that have ONLY ONE ROW in the dataset that corresponds to it.


In [156]:
movie_counts = df["Name"].value_counts()

In [157]:
unique_movie = movie_counts[movie_counts == 1].index.tolist()
print(unique_movie)

['The Covenant', 'Blue Mountain State', 'Shiva Baby', 'Seed of Chucky', 'Fracture', 'Osmosis Jones', 'Slap Shot', 'All Good Things', 'Eden Lake', 'Bound', 'Catch the Bullet', 'xXx', 'The Big Sick', 'Robots', 'North by Northwest', 'Last Week Tonight with John Oliver', 'La montaña sagrada', 'Tears of the Sun', 'Mission: Impossible III', 'Det som göms i snö', 'Hell Fest', 'Future Man', 'Master of None', 'Gifted', 'Anchorman 2: The Legend Continues', 'We Can Be Heroes', 'Triple Frontier', "The King's Speech", '2 Fast 2 Furious', 'Pushing Daisies', 'Jigsaw', 'Pitch Perfect 3', 'The Fighter', 'Orphan: First Kill', 'Green Room', 'Elizabeth', 'The Shallows', 'Family Ties', 'Life in Pieces', 'Hill Street Blues', 'The Legend of Tarzan', 'Moonshine', 'Father of the Bride Part 3 (ish)', 'Humans', 'Friday the 13th Part VI: Jason Lives', 'Operation Fortune: Ruse de guerre', 'Bend It Like Beckham', 'Friday the 13th: A New Beginning', 'Witness for the Prosecution', 'The Dark Tower', 'A Good Year', 'Br

#### 1.3.3. Dropping Movies with Duplicate Names

Finally, use this `Name` list from 1.3.2 to drop all rows that correspond to a `Name` that had more than one row assigned to it. (Or in other words drop the movies that we considered "duplicates").

How many additional rows did you drop?

In [158]:
nonunique_movie = movie_counts[movie_counts >= 2]
nonunique_movie

Name
King Kong                                         5
The Mummy                                         4
Get Shorty                                        4
The Saint                                         4
Lost in Space                                     4
                                                 ..
Bad Girls                                         2
Tristan + Isolde                                  2
Night at the Museum: Battle of the Smithsonian    2
13 Minutes (II)                                   2
Stepmom                                           2
Name: count, Length: 1283, dtype: int64

In [117]:
df_drop_duplicates = df.copy()
na_values = nonunique_movie
na_values

Name
King Kong                                         5
The Mummy                                         4
Get Shorty                                        4
The Saint                                         4
Lost in Space                                     4
                                                 ..
Bad Girls                                         2
Tristan + Isolde                                  2
Night at the Museum: Battle of the Smithsonian    2
13 Minutes (II)                                   2
Stepmom                                           2
Name: count, Length: 1283, dtype: int64

In [118]:
df_drop_duplicates = df_drop_duplicates.dropna()
df_drop_duplicates

Name  Date  Rate     Votes  \
0                             No Time to Die  2021   7.6  107163.0   
2                  The Many Saints of Newark  2021   6.4   27145.0   
9                              Sex Education  2021   8.3  220201.0   
10                          The Walking Dead  2021   8.2  906327.0   
13                              The Sopranos  2021   9.2  340006.0   
...                                      ...   ...   ...       ...   
6169               The Red Sea Diving Resort  2019   6.6   26953.0   
6171                              Chugyeokja  2008   7.8   62423.0   
6173  The Human Centipede II (Full Sequence)  2011   3.8   37492.0   
6175      Before the Devil Knows You're Dead  2007   7.3  100668.0   
6177                              Death Race  2008   6.3  203578.0   

                            Genre  Duration    Type Certificate    Nudity  \
0     Action, Adventure, Thriller     163.0    Film       PG-13      Mild   
2                    Crime, Drama     120.0    Film           R  Moderate   
9                   Comedy, Drama      45.0  Series       TV-MA    Severe   
10        Drama, Horror, Thriller      44.0  Series       TV-14      Mild   
13                   Crime, Drama      55.0  Series       TV-MA    Severe   
...                           ...       ...     ...         ...       ...   
6169     Drama, History, Thriller     129.0    Film       TV-MA      Mild   
6171      Action, Crime, Thriller     125.0    Film   Not Rated      Mild   
6173                       Horror      91.0    Film   Not Rated    Severe   
6175       Crime, Drama, Thriller     117.0    Film           R    Severe   
6177     Action, Sci-Fi, Thriller     105.0    Film           R      Mild   

      Violence Profanity   Alcohol Frightening  
0     Moderate      Mild      Mild    Moderate  
2       Severe    Severe  Moderate    Moderate  
9         Mild    Severe  Moderate        Mild  
10      Severe  Moderate      Mild      Severe  
13      Severe    Severe    Severe      Severe  
...        ...       ...       ...         ...  
6169  Moderate  Moderate      Mild      Severe  
6171    Severe  Moderate      Mild      Severe  
6173    Severe    Severe      Mild      Severe  
6175  Moderate    Severe    Severe      Severe  
6177    Severe    Severe      Mild    Moderate  

[2713 rows x 13 columns]

## 2. ReelView Dataset

### 2.1. Reading the csv

1. Read the `ReelView.csv` into a dataframe. The strings that are used to represent missing values in this csv file are "". So your pd.read_csv() function should automatically convert these values to NaN without you having to specify that "" is a missing value string.
2. Show the first 5 rows.
3. Show how many rows this dataframe has.

In [164]:
df_reel = pd.read_csv('ReelView.csv', na_values=[""])
df_reel.head()
#come back to

title  release_date                director  \
0  Terms of Endearment        1983.0         James L. Brooks   
1                 1917        2019.0              Sam Mendes   
2    Spies in Disguise        2019.0  Nick Bruno, Troy Quane   
3           Just Mercy        2019.0   Destin Daniel Cretton   
4         Little Women        2019.0            Greta Gerwig   

                                                cast  
0  Shirley MacLaine, Debra Winger, Jack Nicholson...  
1                George MacKay, Dean-Charles Chapman  
2  Will Smith, Tom Holland, Rashida Jones, Ben Me...  
3  Michael B. Jordan, Jamie Foxx, Brie Larson, Ti...  
4  Saoirse Ronan, Chris Cooper, Louis Garrel, Jam...

In [165]:
df_reel.shape[0]

3990

### 2.2. Missing Value Cleaning
1. First create a dataframe that only contains the the columns that we care about for this pre-analsys: `title`, `release_date`, and `cast`.
2. Then drop all rows with NaN values from this dataframe.
3. How many rows did we drop?

In [124]:
df_reel3 = df_reel[["title","release_date","cast"]]
df_reel3

title  release_date  \
0      Terms of Endearment        1983.0   
1                     1917        2019.0   
2        Spies in Disguise        2019.0   
3               Just Mercy        2019.0   
4             Little Women        2019.0   
...                    ...           ...   
3985  Music of Chance, The        1993.0   
3986           Ultraviolet        2006.0   
3987               Speed 2        1997.0   
3988               Ex, The        2007.0   
3989                 Venus        2006.0   

                                                   cast  
0     Shirley MacLaine, Debra Winger, Jack Nicholson...  
1                   George MacKay, Dean-Charles Chapman  
2     Will Smith, Tom Holland, Rashida Jones, Ben Me...  
3     Michael B. Jordan, Jamie Foxx, Brie Larson, Ti...  
4     Saoirse Ronan, Chris Cooper, Louis Garrel, Jam...  
...                                                 ...  
3985  Mandy Patinkin, James Spader, M. Emmet Walsh, ...  
3986  Milla Jovovich, Cameron Bright, Nick Chinlund,...  
3987  Sandra Bullock, Jason Patric, Willem Dafoe, Te...  
3988  Zach Braff, Amanda Peet, Jason Bateman, Charle...  
3989  Peter O'Toole, Jodie Whittaker, Leslie Phillip...  

[3990 rows x 3 columns]

In [125]:
df_reel3.dropna()

title  release_date  \
0      Terms of Endearment        1983.0   
1                     1917        2019.0   
2        Spies in Disguise        2019.0   
3               Just Mercy        2019.0   
4             Little Women        2019.0   
...                    ...           ...   
3985  Music of Chance, The        1993.0   
3986           Ultraviolet        2006.0   
3987               Speed 2        1997.0   
3988               Ex, The        2007.0   
3989                 Venus        2006.0   

                                                   cast  
0     Shirley MacLaine, Debra Winger, Jack Nicholson...  
1                   George MacKay, Dean-Charles Chapman  
2     Will Smith, Tom Holland, Rashida Jones, Ben Me...  
3     Michael B. Jordan, Jamie Foxx, Brie Larson, Ti...  
4     Saoirse Ronan, Chris Cooper, Louis Garrel, Jam...  
...                                                 ...  
3985  Mandy Patinkin, James Spader, M. Emmet Walsh, ...  
3986  Milla Jovovich, Cameron Bright, Nick Chinlund,...  
3987  Sandra Bullock, Jason Patric, Willem Dafoe, Te...  
3988  Zach Braff, Amanda Peet, Jason Bateman, Charle...  
3989  Peter O'Toole, Jodie Whittaker, Leslie Phillip...  

[3694 rows x 3 columns]

### 2.3. Duplicate Name Cleaning

#### 2.3.1. List of Non-Duplicated Movie Names

Create a list of all movie `title`s in this dataframe that have ONLY ONE ROW in the dataset that corresponds to it.

In [126]:
title_counts = df_reel["title"].value_counts()
title_counts

title
All the King's Men                                        3
Godzilla                                                  3
Lion King, The                                            3
Halloween                                                 3
Hamlet                                                    3
                                                         ..
Pirates of the Caribbean: The Curse of the Black Pearl    1
Peter's Friends                                           1
Mission: Impossible III                                   1
Silver City                                               1
Venus                                                     1
Name: count, Length: 3898, dtype: int64

In [127]:
unique_titles = title_counts[title_counts == 1]
unique_titles

title
Lucky Number Slevin                                       1
Once                                                      1
Stronger                                                  1
Kingsman: The Golden Circle                               1
LEGO Ninjago Movie, The                                   1
                                                         ..
Pirates of the Caribbean: The Curse of the Black Pearl    1
Peter's Friends                                           1
Mission: Impossible III                                   1
Silver City                                               1
Venus                                                     1
Name: count, Length: 3811, dtype: int64

#### 2.3.2. Dropping Movies with Duplicate Names

Finally, use this `title` list from 2.3.1 to drop all rows that correspond to a `title` that had more than one row assigned to it. (Or in other words drop the movies that we considered "duplicates").

How many additional rows did you drop?

In [128]:
nonunique_titles = title_counts[title_counts >= 2]
nonunique_titles

title
All the King's Men      3
Godzilla                3
Lion King, The          3
Halloween               3
Hamlet                  3
                       ..
Thing, The              2
Little Women            2
King Kong               2
Proof                   2
Beauty and the Beast    2
Name: count, Length: 87, dtype: int64

In [129]:
df_reel_copy = df_reel.copy()
na_values = nonunique_titles
df_reel_copy

title  release_date                director  \
0      Terms of Endearment        1983.0         James L. Brooks   
1                     1917        2019.0              Sam Mendes   
2        Spies in Disguise        2019.0  Nick Bruno, Troy Quane   
3               Just Mercy        2019.0   Destin Daniel Cretton   
4             Little Women        2019.0            Greta Gerwig   
...                    ...           ...                     ...   
3985  Music of Chance, The        1993.0             Philip Haas   
3986           Ultraviolet        2006.0             Kurt Wimmer   
3987               Speed 2        1997.0             Jan de Bont   
3988               Ex, The        2007.0            Jesse Peretz   
3989                 Venus        2006.0           Roger Michell   

                                                   cast  
0     Shirley MacLaine, Debra Winger, Jack Nicholson...  
1                   George MacKay, Dean-Charles Chapman  
2     Will Smith, Tom Holland, Rashida Jones, Ben Me...  
3     Michael B. Jordan, Jamie Foxx, Brie Larson, Ti...  
4     Saoirse Ronan, Chris Cooper, Louis Garrel, Jam...  
...                                                 ...  
3985  Mandy Patinkin, James Spader, M. Emmet Walsh, ...  
3986  Milla Jovovich, Cameron Bright, Nick Chinlund,...  
3987  Sandra Bullock, Jason Patric, Willem Dafoe, Te...  
3988  Zach Braff, Amanda Peet, Jason Bateman, Charle...  
3989  Peter O'Toole, Jodie Whittaker, Leslie Phillip...  

[3990 rows x 4 columns]

In [130]:
df_reel_copy.dropna()
#check

title  release_date                director  \
0      Terms of Endearment        1983.0         James L. Brooks   
1                     1917        2019.0              Sam Mendes   
2        Spies in Disguise        2019.0  Nick Bruno, Troy Quane   
3               Just Mercy        2019.0   Destin Daniel Cretton   
4             Little Women        2019.0            Greta Gerwig   
...                    ...           ...                     ...   
3985  Music of Chance, The        1993.0             Philip Haas   
3986           Ultraviolet        2006.0             Kurt Wimmer   
3987               Speed 2        1997.0             Jan de Bont   
3988               Ex, The        2007.0            Jesse Peretz   
3989                 Venus        2006.0           Roger Michell   

                                                   cast  
0     Shirley MacLaine, Debra Winger, Jack Nicholson...  
1                   George MacKay, Dean-Charles Chapman  
2     Will Smith, Tom Holland, Rashida Jones, Ben Me...  
3     Michael B. Jordan, Jamie Foxx, Brie Larson, Ti...  
4     Saoirse Ronan, Chris Cooper, Louis Garrel, Jam...  
...                                                 ...  
3985  Mandy Patinkin, James Spader, M. Emmet Walsh, ...  
3986  Milla Jovovich, Cameron Bright, Nick Chinlund,...  
3987  Sandra Bullock, Jason Patric, Willem Dafoe, Te...  
3988  Zach Braff, Amanda Peet, Jason Bateman, Charle...  
3989  Peter O'Toole, Jodie Whittaker, Leslie Phillip...  

[3693 rows x 4 columns]

## 3. Merging the Two Dataframes

We'd next like to merge these two cleaned dataframes on the `Name` (ie. `title`) of the film/series.

### 3.1. Outer Join Merge

First, it's possible that not every `Name` in the ratings dataframe may have a match to every `title` in the actors dataframe, and vice versa. To help us assess how many movies in each dataframe don't have a match, let's first perform an **outer join** merge of the two dataframes on the `Name` (ie. `title`) of the movie/film.

In [131]:
name_title_merge = pd.merge(df,
                 df_reel_copy,
                 left_on=['Name'],
                right_on=['title'],
                 how='outer',
                indicator=True)
name_title_merge

Name    Date  Rate     Votes  \
0                  No Time to Die  2021.0   7.6  107163.0   
1                      The Guilty  2021.0   6.3   64375.0   
2                      The Guilty  2000.0   6.1    3446.0   
3       The Many Saints of Newark  2021.0   6.4   27145.0   
4     Venom: Let There Be Carnage  2021.0   6.4   30443.0   
...                           ...     ...   ...       ...   
8862                          NaN     NaN   NaN       NaN   
8863                          NaN     NaN   NaN       NaN   
8864                          NaN     NaN   NaN       NaN   
8865                          NaN     NaN   NaN       NaN   
8866                          NaN     NaN   NaN       NaN   

                            Genre  Duration  Type Certificate    Nudity  \
0     Action, Adventure, Thriller     163.0  Film       PG-13      Mild   
1          Crime, Drama, Thriller      90.0  Film           R       NaN   
2          Crime, Drama, Thriller     108.0  Film           R  Moderate   
3                    Crime, Drama     120.0  Film           R  Moderate   
4       Action, Adventure, Sci-Fi      97.0  Film       PG-13       NaN   
...                           ...       ...   ...         ...       ...   
8862                          NaN       NaN   NaN         NaN       NaN   
8863                          NaN       NaN   NaN         NaN       NaN   
8864                          NaN       NaN   NaN         NaN       NaN   
8865                          NaN       NaN   NaN         NaN       NaN   
8866                          NaN       NaN   NaN         NaN       NaN   

      Violence Profanity   Alcohol Frightening                        title  \
0     Moderate      Mild      Mild    Moderate               No Time to Die   
1          NaN    Severe       NaN    Moderate                          NaN   
2          NaN      Mild       NaN         NaN                          NaN   
3       Severe    Severe  Moderate    Moderate                          NaN   
4     Moderate  Moderate      Mild    Moderate  Venom: Let There Be Carnage   
...        ...       ...       ...         ...                          ...   
8862       NaN       NaN       NaN         NaN         Music of Chance, The   
8863       NaN       NaN       NaN         NaN                  Ultraviolet   
8864       NaN       NaN       NaN         NaN                      Speed 2   
8865       NaN       NaN       NaN         NaN                      Ex, The   
8866       NaN       NaN       NaN         NaN                        Venus   

      release_date            director  \
0           2021.0  Cary Joji Fukunaga   
1              NaN                 NaN   
2              NaN                 NaN   
3              NaN                 NaN   
4           2021.0         Andy Serkis   
...            ...                 ...   
8862        1993.0         Philip Haas   
8863        2006.0         Kurt Wimmer   
8864        1997.0         Jan de Bont   
8865        2007.0        Jesse Peretz   
8866        2006.0       Roger Michell   

                                                   cast      _merge  
0     Daniel Craig, Jeffrey Wright, Lea Seydoux, Ram...        both  
1                                                   NaN   left_only  
2                                                   NaN   left_only  
3                                                   NaN   left_only  
4     Tom Hardy, Woody Harrelson, Michelle Williams,...        both  
...                                                 ...         ...  
8862  Mandy Patinkin, James Spader, M. Emmet Walsh, ...  right_only  
8863  Milla Jovovich, Cameron Bright, Nick Chinlund,...  right_only  
8864  Sandra Bullock, Jason Patric, Willem Dafoe, Te...  right_only  
8865  Zach Braff, Amanda Peet, Jason Bateman, Charle...  right_only  
8866  Peter O'Toole, Jodie Whittaker, Leslie Phillip...  right_only  

[8867 rows x 18 columns]

### 3.2. Rating Dataset `Names` with No Match

First, how many movie `Names` in our ratings dataframe did not have a match with a movie `title` in our actors dataframe?

Hint: The **.isna().sum()** functions might help you.

In [132]:
name_title_merge["Name"].isna().sum()
#Come back to

2621

### 3.3. Actor Dataset `titles` with No Match

Next, how many movie `titles` in our actors dataframe did not have a match with a movie `Name` in our ratings dataframe?

Hint: The **.isna().sum()** functions might help you.

In [166]:
name_title_merge["title"].isna().sum()
#Come back to

4505

### 3.4. Inner Join Merge

Let's next perform an **inner join** merge of ratings dataframe and the actors dataframe on the `Name` (ie. `title`) of the movie/film.

In [134]:
name_title_inner_merge = pd.merge(df,df_reel_copy,
        how='inner',
        left_on=["Name"],
        right_on=["title"])
name_title_inner_merge

Name  Date  Rate     Votes  \
0                                No Time to Die  2021   7.6  107163.0   
1                   Venom: Let There Be Carnage  2021   6.4   30443.0   
2                                      Free Guy  2021   7.3  153835.0   
3                                   Black Widow  2021   6.8  246603.0   
4     Shang-Chi and the Legend of the Ten Rings  2021   7.9  109552.0   
...                                         ...   ...   ...       ...   
1736                           Double Indemnity  1944   8.3  150448.0   
1737         Before the Devil Knows You're Dead  2007   7.3  100668.0   
1738         Before the Devil Knows You're Dead  2007   7.3  100668.0   
1739                                 Death Race  2008   6.3  203578.0   
1740                                 Death Race  2008   6.3  203578.0   

                            Genre  Duration  Type Certificate  Nudity  \
0     Action, Adventure, Thriller     163.0  Film       PG-13    Mild   
1       Action, Adventure, Sci-Fi      97.0  Film       PG-13     NaN   
2       Action, Adventure, Comedy     115.0  Film       PG-13    Mild   
3       Action, Adventure, Sci-Fi     134.0  Film       PG-13     NaN   
4      Action, Adventure, Fantasy     132.0  Film       PG-13     NaN   
...                           ...       ...   ...         ...     ...   
1736      Crime, Drama, Film-Noir     107.0  Film      Passed     NaN   
1737       Crime, Drama, Thriller     117.0  Film           R  Severe   
1738       Crime, Drama, Thriller     117.0  Film           R  Severe   
1739     Action, Sci-Fi, Thriller     105.0  Film           R    Mild   
1740     Action, Sci-Fi, Thriller     105.0  Film           R    Mild   

      Violence Profanity Alcohol Frightening  \
0     Moderate      Mild    Mild    Moderate   
1     Moderate  Moderate    Mild    Moderate   
2     Moderate  Moderate     NaN        Mild   
3     Moderate      Mild    Mild    Moderate   
4     Moderate      Mild    Mild        Mild   
...        ...       ...     ...         ...   
1736      Mild       NaN    Mild        Mild   
1737  Moderate    Severe  Severe      Severe   
1738  Moderate    Severe  Severe      Severe   
1739    Severe    Severe    Mild    Moderate   
1740    Severe    Severe    Mild    Moderate   

                                          title  release_date  \
0                                No Time to Die        2021.0   
1                   Venom: Let There Be Carnage        2021.0   
2                                      Free Guy        2021.0   
3                                   Black Widow        2021.0   
4     Shang-Chi and the Legend of the Ten Rings        2021.0   
...                                         ...           ...   
1736                           Double Indemnity           NaN   
1737         Before the Devil Knows You're Dead        2007.0   
1738         Before the Devil Knows You're Dead        2007.0   
1739                                 Death Race        2008.0   
1740                                 Death Race        2008.0   

                   director                                               cast  
0        Cary Joji Fukunaga  Daniel Craig, Jeffrey Wright, Lea Seydoux, Ram...  
1               Andy Serkis  Tom Hardy, Woody Harrelson, Michelle Williams,...  
2                Shawn Levy  Ryan Reynolds, Jodie Comer, Taika Waititi, Joe...  
3            Cate Shortland  Scarlett Johansson, Florence Pugh, Rachel Weis...  
4     Destin Daniel Cretton  Simu Liu, Awkwafina, Meng’er Zhang, Tony Leung...  
...                     ...                                                ...  
1736           Billy Wilder  Fred MacMurray, Barbara Stanwyck, Edward G. Ro...  
1737           Sidney Lumet  Philip Seymour Hoffman, Ethan Hawke, Albert Fi...  
1738           Sidney Lumet  Philip Seymour Hoffman, Ethan Hawke, Albert Fi...  
1739     Paul W.S. Anderson  Jason Statham, Joan Allen, Ian McShane, Tyrese...  
1740     Paul W.S. Anderson  Jason Stath

### 3.5. Rating and Actor Dataset `Name`(`title`) Matches

How many `Name`(`title`) matches were there in the rating and actor dataset?

In [135]:
name_title_inner_merge.shape[0]

1741

### 3.6. Representativeness of the Merged Dataset

Would you be comfortable stating that this new joined dataset is *representative* of the ratings of all the POPULAR films/series that were listed in our original `imdb.csv` file? Why or why not?

No, this new joined dataset doesn't contain all the 2713 movies in the imdb dataset so it cannot be representative of it.

## 4. Creating Famous Actor Variables

### 4.1. Famous Actor Variable 1: CHOOSE YOUR OWN

Use this merged dataframe from #3.4 to create a new variable that indicates whether or not a famous actor was in the film. You can choose your own famous actor. Make sure that it's an actor in at least one film in this merged dataframe from #3.4.

In [177]:
name_title_inner_merge.dtypes

Name                  object
Date                   int64
Rate                 float64
Votes                float64
Genre                 object
Duration             float64
Type                  object
Certificate           object
Nudity                object
Violence              object
Profanity             object
Alcohol               object
Frightening           object
title                 object
release_date         float64
director              object
cast                  object
has_Margot_Robbie       bool
dtype: object

In [178]:
name_title_inner_merge["has_Margot_Robbie"] = name_title_inner_merge["cast"].apply(lambda x: ('Margot Robbie' in x))
#name_title_inner_merge["has_Margot_Robbie"] = "Margot Robbie".isin(name_title_inner_merge["cast"])#.apply(lambda x:("Margot Robbie" in x))#.apply(lambda x : ('Margot Robbie' in x))
name_title_inner_merge[["Name","has_Margot_Robbie"]]
#type needs to be changed from float to string

TypeError: argument of type 'float' is not iterable

### 4.2. Famous Actor Variable 2: CHOOSE YOUR OWN

Use this merged dataframe from #3.4 to create a new variable that indicates whether or not a famous actor was in the film. You can choose your own famous actor. Make sure that it's an actor in at least one film in this merged dataframe from #3.4.

In [180]:
name_title_inner_merge["has_Ana_de_Armas"] = name_title_inner_merge["cast"].apply(lambda x: ('Ana de Armas' in x))
name_title_inner_merge[["Name","has_Ana_de_Armas"]]

TypeError: argument of type 'float' is not iterable

### 4.3. Actor Rating Comparison

Create two boxplots.
1. One boxplot should show the ratings distribution of films with your first actor in the cast.
2. One boxplot should show the ratings distribution of films with your second actor in the cast.
3. Does one of these actor's movies have a comparatively much higher distribution of ratings than the other actor? Explain.

In [181]:
sns.boxplot(x="has_Margot_Robbie", y="Rate", data=name_title_inner_merge[name_title_inner_merge["has_Margot_Robbie"] == True])
plt.show()

ValueError: min() arg is an empty sequence

In [182]:
sns.boxplot(x="has_Ana_de_Armas", y="Rate", data=name_title_inner_merge[name_title_inner_merge["has_Ana_de_Armas"] == True])
plt.show()

KeyError: 'has_Ana_de_Armas'

### Contribution Report
These contribution reports should be included in all group lab assignments. In this contribution report below you should list of the following:
1. Names of each team member.
2. Contributions of each team member to report.

*For example:*

*<u>Teammates:</u>*

*1. John Smith: did number 1,2 and 3*

*2. Jane Doe: did number 4, and 5*